sbol_utilities/component.py

```diff
@@ -54,6 +54,21 @@ def check_roles(component: sbol3.Component) -> bool:
     return check_roles(obj) and has_dna_type(obj) and len(obj.sequences) == 1
 
 
+def filter_by_roles(doc: sbol3.Document, required_role: str) -> List[sbol3.Identified]:
+    """Search the entire provided document, and return components which have specified role as one of its roles.
+
+    :param doc: sbol3 document to search in
+    :param required_role: the role which must be present in components
+    :return: list of matched components
+    """
+    # callable to be provided as an arg to find_all() of Document class
+    def has_role(obj: sbol3.Identified) -> bool:
+        return isinstance(obj, sbol3.Component) and required_role in obj.roles
+
+    # search through all Identified objects
+    return doc.find_all(has_role)
```

Comment:
I think this can be done more parsimoniously by calling the new `filter_top_level` function that you just added.
While doing this, you might also want to add a complementary `filter_by_type` function.
If both of these return iterables (per `filter_top_level`), then they can be chained together as well.

Reply:
I have tried using `filter_top_level` here, but the `by_roles`, and `by_types` functions are reduced just to a helper function for `filter_top_level`. I am not sure if this is what you meant, can you look at the unit test functions and see if they use these functions in a good way?

Reply:
Having simpler helper functions is fine. As I note below, I think it will do better if they return lambdas, however.